### PR TITLE
Add prometheus exporter to expose validation result in kops-controller

### PR DESCRIPTION
**Pull request by h3poteto**
Ref #9370 

I'm sorry for the large changes.

I add prometheus exporter server in kops-controller. This server collects validation result: ready node, healthy component, unhealthy component, pending pod, unknown pod, not ready pod and missing pod. 
@ari-becker Please tell me if there is something missing from the metrics you want.

- Add `registryPath` and `ListenMetrics` in server options to launch prometheus exporter
- Add collector and exporter for prometheus
- Add a k8s service to connect prometheus exporter

Here is an example
```
# HELP kops_component_healthy Healthy components of the kubernetes cluster.
# TYPE kops_component_healthy gauge
kops_component_healthy{cluster_name="playground.k8s.h3poteto.dev"} 4
# HELP kops_component_unhealthy Unhealthy components of the kubernetes cluster.
# TYPE kops_component_unhealthy gauge
kops_component_unhealthy{cluster_name="playground.k8s.h3poteto.dev"} 0
# HELP kops_node_notReady Not ready nodes in the kubernetes cluster.
# TYPE kops_node_notReady gauge
kops_node_notReady{cluster_name="playground.k8s.h3poteto.dev"} 0
# HELP kops_node_ready Ready nodes in the kubernetes cluster.
# TYPE kops_node_ready gauge
kops_node_ready{cluster_name="playground.k8s.h3poteto.dev"} 2
# HELP kops_pod_missing Missing pods in the kubernetes cluster.
# TYPE kops_pod_missing gauge
kops_pod_missing{cluster_name="playground.k8s.h3poteto.dev"} 0
# HELP kops_pod_notReady Not ready pods in the kubernetes cluster.
# TYPE kops_pod_notReady gauge
kops_pod_notReady{cluster_name="playground.k8s.h3poteto.dev"} 0
# HELP kops_pod_pending Pending pods in the kubernetes cluster.
# TYPE kops_pod_pending gauge
kops_pod_pending{cluster_name="playground.k8s.h3poteto.dev"} 0
# HELP kops_pod_unknown Unknown pods in the kubernetes cluster.
# TYPE kops_pod_unknown gauge
kops_pod_unknown{cluster_name="playground.k8s.h3poteto.dev"} 0
```

### CODE REVIEW

File: cmd/kops-controller/pkg/metrics/collector/validate.go
Comment (h3poteto) [inline cmd/kops-controller/pkg/metrics/collector/validate.go:240]: These codes are almost same as https://github.com/h3poteto/kops/blob/master/pkg/validation/validate_cluster.go .
I think that I should use same functions as `validate_cluster,go` , but for that I need to refactor `validate_cluster.go` . If I contain that changes in this pull request, changes too big. So I will refactor this codes after this pull request is merged.

**Comment by ari-becker** [comment]
Amazing @h3poteto !! This will make our lives so much easier.

Looking forward to seeing this merged.

File: cmd/kops-controller/pkg/metrics/collector/validate.go
Comment (johngmyers) [reply cmd/kops-controller/pkg/metrics/collector/validate.go:256]: I'm not a fan of duplicating code like this. If you need to refactor, then do that first. But why can't you just call validate, then classify the returned failures?

**Review by johngmyers — CHANGES_REQUESTED** [review]
A useful metric or metrics would be the count of NotReady nodes, perhaps broken down by instanceGroup. That would probably require a minor refactor to return the `cloudGroups` from cluster validation.

I'm not sure including this in kops-controller is advisable, especially since that has some key secrets in it. Perhaps a new Deployment is advisable.

File: cmd/kops-controller/pkg/metrics/collector/validate.go
Comment (h3poteto) [reply cmd/kops-controller/pkg/metrics/collector/validate.go:256]: I can't refactor it before implement prometheus exporter, because I don't know which methods in `validate_cluster` is required in prometheus exporter. If you don't care about the amount of change, I can refactor it now.

Validator method only returns `ValidationError` struct: https://github.com/kubernetes/kops/blob/55c5c5cf0987a714e2dc7db9fc19b284029c4d61/pkg/validation/validate_cluster.go#L47-L53
It contains only validation error message, so I can't classify it by the required metrics. This struct is designed with the intention of producing errors on stdout, it is not match prometheus metrics.

File: cmd/kops-controller/pkg/config/options.go
Comment (h3poteto) [reply cmd/kops-controller/pkg/config/options.go:55]: Oh, that's true.

File: upup/pkg/fi/cloudup/template_functions.go
Comment (h3poteto) [reply upup/pkg/fi/cloudup/template_functions.go:383]: When is that? Please tell me, because I don't expect such a case.

File: pkg/model/iam/iam_builder.go
Comment (h3poteto) [reply pkg/model/iam/iam_builder.go:786]: I see.

Comment (h3poteto) [reply pkg/model/iam/iam_builder.go]: https://github.com/kubernetes/kops/pull/10175/commits/7283946296f39c83609251128240f52a97cddeee
I fixed.

File: cmd/kops-controller/pkg/config/options.go
Comment (h3poteto) [reply cmd/kops-controller/pkg/config/options.go]: https://github.com/kubernetes/kops/pull/10175/commits/12c8cdf9fdabd821e327650d4116d2f83454c967
I fixed.

**Comment by h3poteto** [comment]
@johngmyers I added a metrics `notReadyNodes` .
https://github.com/kubernetes/kops/pull/10175/commits/41f424c91e6b21212cf3a0433493b7eeca9e1763

**Comment by ari-becker** [comment]
hi @h3poteto , it has been a couple of months - any plans to make further progress here?

**Comment by h3poteto** [comment]
@ari-becker Hi, I'm waiting for review.
And I will rebase this PR, and I will organize commits.

**Comment by olemarkus** [comment]
Looks good to me. I think this one is good for 1.21, but I'd like to see it a bit more in action, so I am more reluctant to cherry pick this to 1.20 as we are entering beta phase this weekend.

/lgtm

**Comment by olemarkus** [comment]
@h3poteto can you add some documentation around this feature?

**Comment by h3poteto** [comment]
@olemarkus I see. Should I add documents in this pull request? Is another pull request good?

**Comment by olemarkus** [comment]
Perhaps best to add to this PR?

**Comment by h3poteto** [comment]
OK, I will add documents.

**Comment by h3poteto** [comment]
@olemarkus I added a document for this feature.
https://github.com/kubernetes/kops/pull/10175/commits/ab6c638780b680ea95aa99c8d59529e2a911a5fc

**Comment by olemarkus** [comment]
This is very valuable. But I was actually thinking of something user facing. Like how to enable it and such.

**Comment by h3poteto** [comment]
I made this change at the request of @ari-becker. Could you explain the problem you are facing, @ari-becker ?

**Comment by h3poteto** [comment]
And this prometheus exporter will be automatically enabled after create a cluster. So user only needs to specify the metrics endpoint to prometheus.

**Comment by ari-becker** [comment]
> I made this change at the request of @ari-becker. Could you explain the problem you are facing, @ari-becker ?

Basically just standard observability concerns - for example, wanting to set up Prometheus alerts if a cluster is unhealthy for too long, putting the metrics in a Grafana dashboard, stuff like that.

**Comment by hakman** [comment]
@h3poteto Would you mind rebasing once more? I will try to look at this over the weekend.

**Comment by olemarkus** [comment]
I am wondering how heavy this is on the AWS API. The functions you are using does a lot of describe* calls, quite a bit of S3 calls.
What happens if you have X clusters in the same accounts, and each of them have 1-2 prometheuses scraping these metrics?

**Comment by h3poteto** [comment]
@olemarkus Eventually, the number of AWS API calls will depends on `scrape_interval` of Prometheus.

> What happens if you have X clusters in the same accounts, and each of them have 1-2 prometheuses scraping these metrics?

In this case, each exporter will call AWS API for each Prometheus request.

**Comment by olemarkus** [comment]
I think this may have to be rethought then. Even normal operations with these functions often hit rate limits. I also doubt people will be able to implement rare enough scraping. Maybe some of this can be solved with caching, but I am concerned users will not understand what scraping actually does under the hood here.

**Comment by h3poteto** [comment]
> Even normal operations with these functions often hit rate limits.

Oh, really. I have never encounter rate limits, so I didn't notice.  This is what you mean by rate limit, right? https://docs.aws.amazon.com/AWSEC2/latest/APIReference/throttling.html

**Comment by hakman** [comment]
@h3poteto If you rebase, should be able to reduce the flakiness of e2e tests.

**Comment by olemarkus** [comment]
> > Even normal operations with these functions often hit rate limits.
> 
> Oh, really. I have never encounter rate limits, so I didn't notice. This is what you mean by rate limit, right? https://docs.aws.amazon.com/AWSEC2/latest/APIReference/throttling.html

Yep. If you have a larger set of clusters, or clusters with many ASGs, you'll hit those from time to time when doing updates and rolling updates. And here you are basically doing both of those operations (dry run) every X seconds.

**Comment by h3poteto** [comment]
Is there anything I can do for the rate limits?
I can only think of introducing cache, but it also complicates the current validation logic.

**Pull request closed without merging**
